api/grpc_server.go

```diff
@@ -321,25 +322,41 @@ func (s SpacemeshGrpcService) SetLoggerLevel(ctx context.Context, msg *pb.SetLog
 
 func (s SpacemeshGrpcService) GetAccountTxs(ctx context.Context, txsSinceLayer *pb.GetTxsSinceLayer) (*pb.AccountTxs, error) {
 	log.Info("GRPC GetAccountTxs msg")
-	acc := types.HexToAddress(txsSinceLayer.Account.Address)
-	if txsSinceLayer.StartLayer > uint64(s.Tx.LatestLayer()) {
+
+	currentPBase := s.Tx.ValidatedLayer()
+
+	addr := types.HexToAddress(txsSinceLayer.Account.Address)
+	minLayer := types.LayerID(txsSinceLayer.StartLayer)
+	if minLayer > s.Tx.LatestLayer() {
 		return &pb.AccountTxs{}, fmt.Errorf("invalid start layer")
 	}
-	var allTxs []types.TransactionId
-	for i := txsSinceLayer.StartLayer; i < uint64(s.Tx.LatestLayer()); i++ {
-		txs := s.Tx.GetTransactionsByDestination(types.LayerID(i), acc)
-		allTxs = append(allTxs, txs...)
-		moreTxs := s.Tx.GetTransactionsByOrigin(types.LayerID(i), acc)
-		allTxs = append(allTxs, moreTxs...)
+
+	txs := pb.AccountTxs{ValidatedLayer: currentPBase.Uint64()}
+
+	blockTxIds := s.getBlockTxIds(minLayer, addr)
+	for _, txId := range blockTxIds {
+		txs.Txs = append(txs.Txs, txId.String())
 	}
 
-	txs := pb.AccountTxs{}
-	for _, x := range allTxs {
-		txs.Txs = append(txs.Txs, x.String())
+	mempoolTxIds := s.TxMempool.GetTxIdsByAddress(addr)
+	for _, txId := range mempoolTxIds {
+		txs.Txs = append(txs.Txs, txId.String())
 	}
+
 	return &txs, nil
 }
 
+func (s SpacemeshGrpcService) getBlockTxIds(minLayer types.LayerID, addr types.Address) []types.TransactionId {
+	var txIds []types.TransactionId
+	for layerId := minLayer; layerId < s.Tx.LatestLayer(); layerId++ {
```

Comment:
please add comments describing this logic

Reply:
I started and it feels like I'm writing the code in pseudocode:

```
for each layer from minLayer to LatestLayer:
  get transactions by destination and by origin and add them to the list
```

Is reading this more efficient than reading the code? If you think it is, I'll add it as a comment.

Reply:
I meant write why we are taking the txs up until `LatestLayer`

Reply:
I think the real question is why we're indexing transactions by layers, at all -- and there's no good answer, IMO.

The superficial answer is: pagination. But the API design doesn't actually do pagination very well. The first time an account is queried (with layer 0) you always get the full list of transactions, no matter how many there are. You can then query transactions only above the last layer you received, but that doesn't actually remove duplication, since transactions that are in the mempool will be included at least twice (at least once when they're in the mempool and another time when they appear in a block), so you still have to dedup in the wallet.

Since we separate the call for transaction details from the call to the list of transactions and the wallet only selectively queries for transaction details, all this "pagination" does is reduce the size of subsequent TX-list query responses, which would be small even if we sent the full list.

Indexing by layer and having this whole `minLayer` and `LatestLayer` mechanism complicates the design way beyond the efficiency gained by slightly reducing the response size or db query length, while introducing possible bugs and edge cases.

Given that this is my opinion on the mechanism, maybe I'm not the best person to explain it 😉

Reply:
ok